pkg/cli/image/manifest/source.go

```diff
@@ -0,0 +1,141 @@
+		if err != nil {
+			return fmt.Errorf("error decoding ImageContentSourcePolicy from %s: %v", o.ImageContentSourcePolicyFile, err)
+		}
+		var icsp *operatorv1alpha1.ImageContentSourcePolicy
+		var ok bool
+		if icsp, ok = icspObj.(*operatorv1alpha1.ImageContentSourcePolicy); !ok {
+			return fmt.Errorf("could not decode ImageContentSourcePolicy from %s", o.ImageContentSourcePolicyFile)
+		}
+		o.ImageContentSourcePolicyList = append(o.ImageContentSourcePolicyList, *icsp)
+	}
+	o.ICSPClientFn = func() (operatorv1alpha1client.ImageContentSourcePolicyInterface, error) {
+		// If ImageContentSourceFile is given, only add ImageContentSource from file, don't search cluster ICSP
+		if len(o.ImageContentSourcePolicyFile) != 0 {
+			return nil, nil
+		}
+		restConfig, err := f.ToRESTConfig()
+		if err != nil {
+			// may or may not be connected to a cluster
+			// don't error if can't connect
+			klog.V(4).Infof("did not connect to an OpenShift 4.x server, will not lookup ImageContentSourcePolicies: %v", err)
+			return nil, nil
+		}
+		icspClient, err := operatorv1alpha1client.NewForConfig(restConfig)
+		if err != nil {
+			// may or may not be connected to a cluster
+			// don't error if can't connect
+			klog.V(4).Infof("did not connect to an OpenShift 4.x server, will not lookup ImageContentSourcePolicies: %v", err)
+			return nil, nil
+		}
+		return icspClient.ImageContentSourcePolicies(), nil
+	}
+	return nil
+}
+
+// GetICSPs will lookup ICSPs from cluster. Since it's not a hard requirement to find ICSPs from cluster, GetICSPs logs errors rather than returning errors.
+func (o *ImageSourceOptions) GetICSPs(icspClient operatorv1alpha1client.ImageContentSourcePolicyInterface) {
+	icsps, err := icspClient.List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		// may or may not have access to ICSPs in cluster
+		// don't error if can't access ICSPs
+		klog.V(4).Infof("did not access any ImageContentSourcePolicies in cluster: %v", err)
+	}
+	if len(icsps.Items) == 0 {
+		klog.V(4).Info("no ImageContentSourcePolicies found in cluster")
+	}
+	o.ImageContentSourcePolicyList = append(o.ImageContentSourcePolicyList, icsps.Items...)
+}
+
+// GetImageSourcePrefixes returns list of 'registry/repository/name' options to try
+// 'registry/repository/name' determined from examining ImageContentSourcePolicy objects.
+// If a valid prefix is obtained from ICSP, return true
+func GetImageSourcePrefixes(icsps []operatorv1alpha1.ImageContentSourcePolicy, image string) (bool, []string, error) {
+	var imageSources []string
+	for _, icsp := range icsps {
+		repoDigestMirrors := icsp.Spec.RepositoryDigestMirrors
+		var sourceMatches bool
+		imageRef, err := reference.Parse(image)
+		if err != nil {
+			return false, imageSources, err
+		}
+		for _, rdm := range repoDigestMirrors {
+			rdmRef, err := reference.Parse(rdm.Source)
+			if err != nil {
+				return false, imageSources, err
+			}
+			if imageRef.AsRepository() == rdmRef.AsRepository() {
```

Comment:
ICSP is used in CRI-O and builds via `registries.conf`, so see their example: https://github.com/containers/image/blob/f1a6f9b50d0286bf208935d4b9ab48117294ac31/registries.conf

If you have the policy
```yaml
source: example.com/foo
mirrors:
- internal-registry-for-example.net/bar
```
then the image `example.com/foo/myimage:latest` should be pulled from `internal-registry-for-example.net/bar/myimage:latest`.

Reply:
> I reviewed only GetImageSourcePrefixes, it doesn't look like a correct implementation. ICSPs define rules for prefixes, not full repository names.
> 
> See also https://github.com/openshift/runtime-utils/blob/c45d4ff3f912/pkg/registries/registries.go#L104, it merges multiple ICSPs together and deals with some edge-cases using topological sorting.

I've added explanation to GetImageSourcePrefixes - I believe the implementation is what is required. From the client side, when extracting from mirrored images, currently this requires access & authorization to the original registry/repository defeating the purpose of mirroring.  This change adds logic to `oc` to try various imageSources (all verified through digests) since `oc` currently _only_ tries the underlying image reference of a mirrored image - for example, if `4.6.0` is mirrored to `localhost:5000/ocp/release`, 
it's reference remains `quay.io/openshift-release-dev/ocp-release@sha256:579cd61047067d8d642dc0bf8b3...` and currently, that is the only location `oc` knows to extract from.  This change allows `oc` to substitute the `registry/repository/name` so the release will be `localhost:5000/openshift/release@sha256:579cd61047067d8d6` and then the extract will then happen from `localhost:5000/openshift/release@sha256:sha-of-some-tool-or-component` that was originally `quay.io/openshift-release-dev/ocp-release@sha256:sha-of-some-tool-or-component`  This substitution/lookup happens in the following order:
*   if --icsp-file is set, then decode/use that - don't look further
*   image reference from user-passed release image (the original un-mirrored source)
*   if ICSP found in cluster, add that to list of image source options, those will be:
      * icsp from cluster
      * user-passed release image (set registry/repo/name from user-given, and use dame digest)

Reply:
I don't doubt it works as you expect it. I'm trying to understand its purpose and assumptions.

It seems you call prefixes strings like `quay.io/openshift-release-dev/ocp-release`. In Docker (and CRI-O/podman/buildah) terminology they are called repositories. CRI-O calls prefixes strings like `quay.io`, `quay.io/openshift-release-dev`, or `quay.io/openshift-release-dev/ocp-release`. So it returns a list of repositories that may contain the image, right?

Do you assume that ICSP is generated by `oc adm release mirror`? If so, it's a strong assumption and it should be clearly documented.

Reply:
right, it returns `repositories` i call them prefixes because repositories is confusing/misleading. I can update to note that `prefix == repositories == registry/repository/name` that makes sense, and yes, this code assumes ICSP is generated by `oc adm release mirror`  but it should not, you are right.  let me look to see if i can remove that assumption

Reply:
I've moved the logic for getting distribution.Repository to library-go see here: https://github.com/openshift/library-go/pull/939